src/tir/schedule/primitive/loop_transformation.cc

```diff
@@ -0,0 +1,420 @@
+  result_srefs.reserve(n);
+  for (size_t i = 0; i < n; i++) {
+    result_srefs.push_back(self->stmt2ref.at(new_stmt.get()));
+    const ForNode* outer_loop = TVM_TYPE_AS(outer_loop, new_stmt, ForNode);
+    new_stmt = outer_loop->body;
+  }
+  return result_srefs;
+}
+
+StmtSRef Fuse(ScheduleState self, const Array<StmtSRef>& loop_srefs) {
+  // Invariance
+  // - The total repeat number has not changed for each direct child block.
+  // - The execution order has not changed. (The block executes with the same
+  //   args and the same order with before.)
+  std::vector<const ForNode*> loops;
+  loops.reserve(loop_srefs.size());
+  StmtSRef outer_loop_sref{nullptr};
+  const ForNode* outer_loop = nullptr;
+  arith::Analyzer analyzer;
+  // Step 1. check correctness
+  for (const StmtSRef& sref : loop_srefs) {
+    const ForNode* loop = TVM_SREF_TO_FOR(loop, sref);
+    if (!loop->annotations.empty() || loop->thread_binding.defined()) {
+      throw HasAnnotationOrThreadBindingError(self->mod, GetRef<For>(loop));
+    }
+    if (outer_loop_sref.defined()) {
+      if (sref->parent != outer_loop_sref.get()) {
+        throw OuterNotInnerParent(self->mod, GetRef<For>(outer_loop), GetRef<For>(loop));
+      }
+      if (!outer_loop->body.same_as(GetRef<For>(loop))) {
+        throw NotOnlyChildError(self->mod, GetRef<For>(outer_loop), GetRef<For>(loop));
+      }
+    }
+    outer_loop_sref = sref;
+    outer_loop = loop;
+    if (!analyzer.CanProve(loop->min == 0)) {
+      throw LoopNotStartWithZeroError(self->mod, GetRef<For>(loop));
+    }
+    loops.push_back(loop);
+  }
+  // Step 2. Create fused loop var and replace the original loop vars
+  std::string suffix;
+  for (size_t i = 1; i < loops.size(); i++) {
+    suffix += "_" + loops[i]->loop_var->name_hint;
+  }
+  suffix += "_fused";
+  Var fused_var = loops[0]->loop_var.copy_with_suffix(suffix);
+  Array<PrimExpr> substitute_value;
+  substitute_value.resize(loops.size());
+  PrimExpr tot = fused_var;
+  for (int i = loops.size() - 1; i >= 0; i--) {
+    substitute_value.Set(i, floormod(tot, loops[i]->extent));
+    tot = floordiv(tot, loops[i]->extent);
+  }
+  Stmt loop_body = loops.back()->body;
+  Map<Block, Block> opaque_block_reuse;
+  auto f_substitute = [&](const Var& v) -> Optional<PrimExpr> {
+    for (size_t i = 0; i < loops.size(); i++) {
+      if (v.same_as(loops[i]->loop_var)) {
+        return substitute_value[i];
+      }
+    }
+    return NullOpt;
+  };
+  Stmt new_stmt =
+      SubstituteVarAndCollectOpaqueBlock(f_substitute, &opaque_block_reuse)(std::move(loop_body));
+  // Step 3. Generate a loop to replace the original loops
+  PrimExpr fused_extent = 1;
+  for (size_t i = 0; i < loops.size(); i++) {
+    fused_extent *= loops[i]->extent;
+  }
+  fused_extent = analyzer.Simplify(fused_extent);
+  new_stmt = For(fused_var, 0, fused_extent, ForKind::kSerial, new_stmt);
+  new_stmt = IterMapSimplifyBlockBinding::SimplifyBindings(new_stmt, GetLoops(loop_srefs[0]),
```

Comment:
Move in to help inplace mutation

```suggestion
  new_stmt = IterMapSimplifyBlockBinding::SimplifyBindings(std::move(new_stmt), GetLoops(loop_srefs[0]),
```